Userland/Demos/Tubes/Tubes.cpp

```diff
@@ -0,0 +1,425 @@
+    glEnable(GL_COLOR_MATERIAL);
+    glColorMaterial(GL_FRONT, GL_AMBIENT_AND_DIFFUSE);
+
+    glEnable(GL_DEPTH_TEST);
+    glDepthFunc(GL_LEQUAL);
+    glEnable(GL_NORMALIZE);
+    glShadeModel(GL_SMOOTH);
+}
+
+void Tubes::timer_event(Core::TimerEvent&)
+{
+    update_tubes();
+    m_gl_context->present();
+    repaint();
+}
+
+void Tubes::update_tubes()
+{
+    if (++m_ticks % reset_every_ticks == 0)
+        reset_tubes();
+
+    double const primitive_size = 2.; // our tubes and spheres are 1 in diameter, so object size is 2
+    double const grid_width = 2.;
+    double const grid_scale = 1. / grid_resolution;
+    double const primitive_scale = 1. / primitive_size;
+    double const tube_length_scale = tube_movement_per_tick * primitive_size;
+    double const tube_thickness_scale = tube_relative_thickness * primitive_scale;
+    for (auto& tube : m_tubes) {
+        if (!tube.active)
+            continue;
+
+        glColor3d(tube.color.x(), tube.color.y(), tube.color.z());
+        glPushMatrix();
+
+        auto pos = tube.position;
+        glTranslated(
+            pos.x() * grid_scale * grid_width - (grid_width / 2.),
+            pos.y() * grid_scale * grid_width - (grid_width / 2.),
+            pos.z() * grid_scale * grid_width - (grid_width / 2.));
+        glScaled(grid_scale, grid_scale, grid_scale);
+
+        // Draw sphere if we're at the start or a corner
+        if (tube.progress_to_target == 0.) {
+            glPushMatrix();
+            glScaled(tube_thickness_scale, tube_thickness_scale, tube_thickness_scale);
+            draw_sphere();
+            glPopMatrix();
+        }
+
+        // Draw tube at the current position
+        glPushMatrix();
+        auto direction_vector = vector_for_direction(tube.direction);
+        auto distance_to_target = (tube.target_position - tube.position).length<double>();
+        auto movement_magnitude = tube.progress_to_target * (distance_to_target - tube_movement_per_tick) / distance_to_target * grid_width;
+        glTranslated(
+            direction_vector.x() * movement_magnitude,
+            direction_vector.y() * movement_magnitude,
+            direction_vector.z() * movement_magnitude);
+        auto tube_rotation = tube_rotation_for_direction(tube.direction);
+        glRotated(tube_rotation.w(), tube_rotation.x(), tube_rotation.y(), tube_rotation.z());
+        glScaled(tube_thickness_scale, tube_thickness_scale, primitive_scale * tube_length_scale);
+
+        draw_tube();
+        glPopMatrix();
+
+        // Move towards target
+        if (tube.progress_to_target >= distance_to_target) {
+            tube.position = tube.target_position;
+            choose_new_direction_for_tube(tube);
+        } else {
+            tube.progress_to_target = min(tube.progress_to_target + tube_movement_per_tick, distance_to_target);
+        }
+
+        glPopMatrix();
+    }
+}
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
```

Comment:
Can we move this into a `main.cpp` please? I realise not all of the demos have one, but it's a weird code structure.

Reply:
Yeah 100% agreed - fixed. The base was copied from Starfield which caused this abnormal structure.